common/_common_soc/1bitsquared_icebreaker/ld/linker.ld

```diff
@@ -8,56 +8,56 @@ INCLUDE regions.ld
 
 SECTIONS
 {
-	.reserved :
-	{
-		/* Space reserved for gateware, shouldn't step on it. */
-		. += 0x30000;
-	} > spiflash
+    /* .data first to take precedence for explicit code we want in SRAM. */
+    .data : AT (_erodata) /* Loaded in spiflash but runtime copied to SRAM. */
+    {
+        . = ALIGN(4);
+        _fdata = .;  /* Start of .data in SRAM, copied to in crt0. */
 
-	.text :
-	{
-		_ftext = .;
-		*(.text.start)
-		*(.text .stub .text.* .gnu.linkonce.t.*)
-		_etext = .;
-	} > spiflash
+        *(.ramtext .ramtext*)
 
-	.rodata :
-	{
-		. = ALIGN(8);
-		_frodata = .;
-		*(.rodata .rodata.* .gnu.linkonce.r.*)
-		*(.rodata1)
-		*(.srodata .srodata.*)
-		. = ALIGN(8);
-		_erodata = .;
-	} > spiflash
+        /* 
+         * We need to ensure .data starts at > 0, otherwise a pointer to the
+         * first element of .data will be mistaken for a nullptr.
+         */
+        . = MAX(., 0x4); 
+
+        *(.data .data*)
+        *(.sdata .sdata*)
+
+        . = ALIGN(4);
+        _edata = .;  /* End of .data in SRAM, end of copying in crt0. */
+    } > sram
 
-	.data : AT (ADDR(.rodata) + SIZEOF (.rodata))
+    /* Start .text at 4 byte boundary after gateware. */
+    _ftext = (ORIGIN(spiflash) + 0x30000 + 3) & ~3;
+    .text _ftext :
 	{
-		. = ALIGN(8);
-		_fdata = .;
-		*(.data .data.* .gnu.linkonce.d.*)
-		*(.data1)
-		*(.ramtext .ramtext.*)
-		_gp = ALIGN(16);
-		*(.sdata .sdata.* .gnu.linkonce.s.* .sdata2 .sdata2.*)
-		_edata = ALIGN(16); /* Make sure _edata is >= _gp. */
-	} > sram
+        *(.text.start)
+        *(.text .text*)
+    } > spiflash
+
+    .rodata :
+    {
+        *(.rodata .rodata*)
+        *(.srodata .srodata*)
+
+        . = ALIGN(4);
+        _erodata = .;  /* Start of .data in spiflash, copied from in crt0. */
+    } > spiflash
 
-	.bss : AT (ADDR(.data) + SIZEOF (.data))
+	.bss :
```

Comment:
The indentation for .bss is a little wonky (it was before as well) -- can we fix it here and below?

Reply:
Great catch! Forgot to change over those tabs to spaces when I was modifying the linker file. Fix pushed :).